### PR TITLE
Update Weights & Biases Integration + Add Prediction Visualization for Object Detection

### PR DESCRIPTION
**Pull request by soumik12345**
# Changes made in this PR

## Log inference results to Weights & Biases

Fixes https://github.com/Deci-AI/super-gradients/issues/953

Optionally log inference results for object detection in [Weights & Biases](https://wandb.ai/site) dashboard using [bounding-box image overlays](https://docs.wandb.ai/guides/track/log/media#image-overlays).

This PR adds an optional parameter `visualize_on_wandb` to `ImagesDetectionPrediction.show()` which when set to `True` would log the predictions using the [bounding-box image overlays](https://docs.wandb.ai/guides/track/log/media#image-overlays) in the [Weights & Biases](https://wandb.ai/site) dashboard of the user.

Here's a sample code:

```python
from super_gradients.training import models
from super_gradients.common.object_names import Models

import wandb

# Initialize Weights & Biases run
wandb.init(project="yolo-nas", entity="geekyrakshit")

net = models.get(Models.YOLO_NAS_S, pretrained_weights="coco")
prediction = net.predict([
    "https://cdn.autonomous.ai/static/upload/images/common/upload/20210919/5-Desk-Setup-with-a-Laptop--Monitor-Ideas-for-2021_10f89201411b.jpg",
    "https://p.imgci.com/db/PICTURES/CMS/352000/352019.jpg",
    "https://p.imgci.com/db/PICTURES/CMS/352200/352298.jpg",
    "https://www.encirclephotos.com/wp-content/uploads/Laos-Ban-Pak-Ou-Man-Riding-Asian-Elephant-1440x961.jpg"
])
prediction.show(visualize_on_wandb=True)
```

This would log the images to be visualized in a media panel in a Weights & Biases dashboard.

👇 Here's how the results would look
[](https://wandb.ai/geekyrakshit/yolo-nas-integration/runs/lz9mjhwy)

Here's a video of how to interact with the bounding-box overlay interface on Weights & Biases 👇

https://github.com/Deci-AI/super-gradients/assets/19887676/fb027d35-adc1-4e93-9d5a-2ea7b32c50f7

## Update `WandBSGLogger`

Fixes https://github.com/Deci-AI/super-gradients/issues/980
Fixes https://github.com/Deci-AI/super-gradients/issues/1008

- `WandBSGLogger` will now call `wandb.init()` only if it has not been called earlier.
- Save model checkpoints versioned with Weights & Biases checkpoints.
- Add an additional parameter `sync_tensorboard` to `WandBSGLogger` that automatically syncs TensorBoard with the wandb run.

👇 Here's a sample training code that works with the changes made in this PR
[](https://colab.research.google.com/drive/1478GGwddwiyWJzb6umfVH5bNl8unx0UR?usp=sharing)

👇 Here's a sample Weights & Biases Run demonstrating the features
[](https://wandb.ai/geekyrakshit/yolo-nas-integration/runs/n1trro70)

👇 Tensorboard instance synced with Weights & Biases Run
[](https://wandb.ai/geekyrakshit/yolo-nas-integration/runs/n1trro70/tensorboard)

👇 Lineage of Model Checkpoints
[](https://wandb.ai/geekyrakshit/yolo-nas-integration/artifacts/model/n1trro70-checkpoint/27fc3b1a6c611bd3dc70/lineage)

### CODE REVIEW

**Comment by soumik12345** [comment]
Hi @BloodAxe 
Can you please take a look at this PR?

File: src/super_gradients/wandb/artifacts.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import os
+import wandb
```
Comment (ofrimasad) [inline src/super_gradients/wandb/artifacts.py:2]: Please use the same import logic we used in the WandbLogger 
```
try:
    import wandb
except (ModuleNotFoundError, ImportError, NameError):
    pass  # no action or logging - this is normal in most cases
```
then check if `wandb is not None`

File: src/super_gradients/wandb/artifacts.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+
+def save_artifact(path):
```
Comment (ofrimasad) [inline src/super_gradients/wandb/artifacts.py:5]: If you define the type to be model, please name the function `save_wandb_model_artifact`
otherwise, call it `save_wandb_artifact(path: str, type: str)`.

note the added explicit typing

Comment (ofrimasad) [reply src/super_gradients/wandb/artifacts.py:5]: please add documentation:
something like:
```
 upload and wandb Artifact.
Note that this function can be called only after wandb.init()
 :param path: the local full path to the pth file to be uploaded 
 ```

File: src/super_gradients/wandb/log_predictions.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import wandb
```
Comment (ofrimasad) [inline src/super_gradients/wandb/log_predictions.py:1]: same here regarding the import please

File: src/super_gradients/common/sg_loggers/wandb_sg_logger.py [hunk 4]
```diff
@@ -102,7 +105,11 @@ def __init__(
                     )
                 wandb_id = self._get_wandb_id()
 
-        run = wandb.init(project=project_name, name=experiment_name, entity=entity, resume=resumed, id=wandb_id, **kwargs)
+        run = (
+            wandb.init(project=project_name, name=experiment_name, entity=entity, resume=resumed, id=wandb_id, sync_tensorboard=sync_tensorboard, **kwargs)
```
Comment (ofrimasad) [inline src/super_gradients/common/sg_loggers/wandb_sg_logger.py:109]: why do we need to synch tensorboard if all information is logged to wandb? Wouldn't this create duplicated data?

File: src/super_gradients/common/sg_loggers/wandb_sg_logger.py [hunk 5]
```diff
@@ -102,7 +105,11 @@ def __init__(
...
+            if wandb.run is None
```
Comment (ofrimasad) [inline src/super_gradients/common/sg_loggers/wandb_sg_logger.py:110]: This is breaking the resume experiment feature.
Moreover, if the run was initialized from outside, we have no way of knowing it is logging into the correct experiment. 
Why would you want to init wandb from outside of the logger?

File: src/super_gradients/common/sg_loggers/wandb_sg_logger.py [hunk 6]
```diff
@@ -246,7 +253,7 @@ def add_checkpoint(self, tag: str, state_dict: dict, global_step: int = 0):
         if self.save_checkpoints_wandb:
             if self.s3_location_available:
                 self.model_checkpoints_data_interface.save_remote_checkpoints_file(self.experiment_name, self._local_dir, name)
-            wandb.save(glob_str=path, base_path=self._local_dir, policy="now")
+            save_artifact(path)
```
Comment (ofrimasad) [inline src/super_gradients/common/sg_loggers/wandb_sg_logger.py:256]: will this have the same effect? 
currently all files are saved inside the wandb experiment. Will the artifact be saved in the same way?

File: src/super_gradients/wandb/log_predictions.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+from super_gradients.training.models.prediction_results import ImageDetectionPrediction, ImagesDetectionPrediction
+
+
+def visualize_image_detection_prediction_on_wandb(prediction: ImageDetectionPrediction, show_confidence: bool):
```
Comment (ofrimasad) [inline src/super_gradients/wandb/log_predictions.py:6]: please make this function private (start with _)

File: src/super_gradients/wandb/log_predictions.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+            box["scores"] = {"confidence": float(round(prediction.prediction.confidence[pred_i], 2))}
+        boxes.append(box)
+
+    wandb_image = wandb.Image(image, boxes={"predictions": {"box_data": boxes, "class_labels": class_id_to_labels}})
+
+    wandb.log({"Predictions": wandb_image})
+
+
+def log_detection_results_to_wandb(prediction: ImagesDetectionPrediction, show_confidence: bool = True):
+    """Log predictions for object detection to Weights & Biases using interactive bounding box overlays.
```
Comment (ofrimasad) [inline src/super_gradients/wandb/log_predictions.py:34]: If I am not mistaken, this has to be called after `wandb.init()` please state this in the documentation and add an if-raise statement

File: src/super_gradients/wandb/artifacts.py (hunk 2 above)
Comment (ofrimasad) [reply src/super_gradients/wandb/artifacts.py:5]: looks like this code is used only inside the `WandBSGLogger`
why not include this as a private function of `WandBSGLogger`?

File: src/super_gradients/common/sg_loggers/wandb_sg_logger.py (hunk 6 above)
Comment (soumik12345) [reply src/super_gradients/common/sg_loggers/wandb_sg_logger.py]: Hi @ofrimasad 
You can refer to https://github.com/Deci-AI/super-gradients/pull/1010#issuecomment-1561244431 regarding this.

File: src/super_gradients/common/sg_loggers/wandb_sg_logger.py (hunk 5 above)
Comment (ofrimasad) [reply src/super_gradients/common/sg_loggers/wandb_sg_logger.py]: Assume I trained my model fro 10 epochs and then stopped it for some reason. now I want to resume the training.
I can just use the `resume` flag in the training params and the wandb run will be connected to the right run ID.
but if I init the run from outside, how would I do this? I need to go to wandb, check the run Id, init the run with the correct run Id and the resume flag. If I just call wandb.init() without the id, the run will not resume.

File: src/super_gradients/common/sg_loggers/wandb_sg_logger.py (hunk 6 above)
Comment (ofrimasad) [reply src/super_gradients/common/sg_loggers/wandb_sg_logger.py]: After testing this, I can see two issues with this change:

1. it dramatically inflates the storage required by the user. Instead of saving only the last version of the model, it saves *all* the checkpoints. Users that are using an on-prem wandb server (like we do) will see a huge increase in storage usage.  

2. you are uploading different files to the same Artifact. During the training, we save ckpt_best.pth, ckpt_latest.pth, ckpt_average.pth. and this code maps all of them to the same Artifact (as different versions)  /

Please add a parameter (in the init), so the users can choose if they want the files to be saved as Artifact or not. 
please differentiate between files with different names.

**Review by ofrimasad — COMMENTED** [review]
I see that you still kept the option to init wandb from outside the logger.
In that case please add a warning that will be presented if the run was initialized from outside the logger. 
"The wandb run was initialized from outside of the WandbSGLogger. This means that SG cannot control the run ID to which this session will be logged. The current run id is: {run.id}"

Thanks

**Review by ofrimasad — COMMENTED** [review]
please use the common logging behavior of SG

**Comment by soumik12345** [comment]
Hi @ofrimasad is there any more changes we need to make as part of this PR?
